### ISSUE TITLE
question: Is MetadataDocument parsed once and cached?

### ISSUE THREAD

**Issue by rasikas** [object/Simple.OData.Client#249]
I'm using this library very extensively and it is working very well. Recently I wanted to assign the MetadataDocument from local file instead of letting library to request it and cache in the very first call.

So whenever we call particular endpoint, I create the setting object and assign the MetadataDocument property. I would like to know whether this is only parsed once no matter how many times we assigned it into a cache (because we create client object for every web request).

or else is there a method to retrieve all cached metadata documents so that we could  check it and avoid assigning if it is already in the cache.

p.s. I want to tag this as a question, but I dont see any labels.

**Comment by object** [comment]
Yes, the metadata document is parsed only once per service URL, whether it's loaded locally or obtained from the actual service.

**Comment by rasikas** [comment]
Thanks a lot for your quick reply

**Comment by benbenbenbenbenben** [comment]
We are creating an OData client per request and we have noticed that OData is parsing our metadata string every time we do this. I would argue the above answer is incorrect.

**Comment by object** [comment]
Are you sure? Can you provide a code example that reproduces it? This shouldn't happen. I ran a test with sessions created multiple times and the metadata string was parsed only once.

**Comment by benbenbenbenbenben** [comment]
Sorry I can't share the code because it's at an onsite, but I can tell you what we're doing.

1. Load metadata into string (outside Simple.OData)
2. Create client using settings and above string set as metadata document

Each time we create a client, `ParseMetadata` from `AdapterFactory` is called and in our case adding 2s~ to execution time. Hence why we needed to change it.

Our app has WebAPI middleware between browser and an OData backend. We only found out after profiling with R#. There is a very iffy hack to quickly solve the problem for us in the changeset I commited to the fork under my user.

[https://github.com/benbenbenbenbenben/Simple.OData.Client/commit/688009a10404ead723e6db400768d39d0453afba](url)

**Comment by sixlettervariables** [comment]
@object we see this with our app. Each ODataClient instance retrieves and parses its own metadata.

Cold start:
```
[2017-03-20T16:09:30.9596965-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:30.9786984-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:30.9866992-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:30.9926998-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:30.9967002-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:31.0177023-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:31.0447050-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:31.0537059-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3020307-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3090314-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3410346-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3550360-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3720377-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.3800385-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:09:34.4060411-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
```
Warm start:
```
[2017-03-20T16:11:43.0969089-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1119104-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1189111-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1439136-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1499142-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1669159-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1789171-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.1839176-04:00][OData] GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.3329325-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.3619354-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.3829375-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.4109403-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.3619354-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.4259418-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.4759468-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
[2017-03-20T16:11:43.7079700-04:00][OData] Request completed: OK / GET request: http://.../OData/$metadata
```
Code which kicks this off looks roughly like this (can't show full details):
```csharp
        public Task InitializeAsync(CancellationToken token)
        {
            return Task.WhenAny(
                Task.WhenAll(
                    this.Get1(token),
                    this.Get2(token),
                    this.Get3(token),
                    this.Get4(token),
                    this.Get5(token),
                    this.Get6(token),
                    this.Get7(token),
                    this.Get8(token)),
                Task.Delay(TimeSpan.FromMinutes(2), token));
        }

        // ...

        public async Task<IEnumerable<Foo>> Get1(CancellationToken token)
        {
            if (cachedFoos == null)
            {
                var items = await this.odataClient.For<Foo>()
                                               .Expand(foo => new { foo.Bar, rsm.Baz})
                                               .FindEntriesAsync(token)
                                               .ConfigureAwait(false);
                cachedFoos = items.ToList().AsReadOnly();
            }

            return cachedFoos;
        }

        // ...

            this.odataClient = new ODataClient(new ODataClientSettings
            {
                Credentials = this.configuration.Credentials,
                BaseUri = this.configuration.EndPointResolver.GetServiceEndPoint(nameof(ODataApi)),
                RenewHttpConnection = false,
#if TRACE
                OnTrace = (format, args) =>
                {
                    if (ODataApiTracing.Enabled)
                    {
                        var msg = String.Format(format, args);
                        if (msg == "Request completed: OK")
                        {
                            Trace.WriteLine($"[{DateTimeOffset.Now:O}][OData] " + msg + " / " + lastMessage);
                        }
                        else
                        {
                            Trace.WriteLine($"[{DateTimeOffset.Now:O}][OData] " + msg);
                            lastMessage = msg; // TODO: move to support Task
                        }
                    }
                },
#endif
            });

```

**Comment by phatcher** [comment]
@sixlettervariables Have a look at my version, I can't easily stop multiple parsing of the metadata, but it does now only retrieve the metadata once - see FindEntryNuGetV2MultiThread()

@object Can you check that your happy for me to merge this, it does pass all the tests, but I had to move stuff around a bit to make it work nicely; still needs a little bit of tidy up to remove obsolete methods but works fine,

**Comment by sixlettervariables** [comment]
@phatcher interesting work, I'm not certain if I have the time to test this internally right now (might be a week or so).

A few notes: You'll want to change `lock(this)` to lock on a private variable (static if need be). Likewise with `lock(MetadataCache)`, as both have no control over who else is locking the lock object. This probably means pushing most of that logic into MetadataCache. The GetAwaiter() and Task.WaitAll inside a lock region probably need to see a revision to avoid that as well. I also noticed an undisposed XmlReader instance.

**Comment by sixlettervariables** [comment]
The other bit I noticed is this (and really any other change like this) ends up with the first request's cancellation token source winning. I'm not really sure how to approach that bit. Two options: separate metadata timeout, or settings "owns" each cache instance and you share settings among clients who should share a cache.

**Comment by sixlettervariables** [comment]
@phatcher I refactored your branch to remove the GetAwaiter/WaitAll inside the lock. I think this accomplishes everything required. (only one metadata request per URI, only one adapter load per metadata)

**Comment by sixlettervariables** [comment]
It appears feasible to remove the multiple requests, but going deeper, because `ISession` is pushed down to the adapter (which loads the model), you would need to add another layer of indirection to avoid the double parsing.

**Comment by object** [comment]
Thank you for collaboration, guys. I just merged the PR.

**Comment by object** [comment]
Released in 4.28.
